bin/python_scripts/find_invalid_tags.py

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Set up environment from ckan.ini
+# export POSTGRES_URL=<sqlalchemy.url from ckan.ini>
+#
+# Execute script like this -
+# python find_invalid_tags.py
+#
+
+import io
+import os
+import sys
+import psycopg2
+import subprocess
+
+import logging
+
+from ckan.logic.schema import default_tags_schema
+from ckan.lib.navl.dictization_functions import validate, Invalid
+
+POSTGRES_URL = os.environ.get('POSTGRES_URL')
+
+logger = logging.getLogger(__name__)
+connection = psycopg2.connect(POSTGRES_URL)
+
+
+def setup_logging():
+    _format = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')
+    logger.setLevel(logging.INFO)
+    c_handler = logging.StreamHandler()
+    c_handler.setFormatter(_format)
+    logger.addHandler(c_handler)
+
+    f_handler = logging.FileHandler('find_invalid_tags.log')
+    f_handler.setFormatter(_format)
+    logger.addHandler(f_handler)
+
+    logger.info('====================================================================')
+
+
+# only target packages with tags that are active
+def get_all_tags():
+    cursor = connection.cursor()
+    sql = """
+    SELECT tag.name, count(tag.name) FROM tag, package, package_tag
+    WHERE tag.id = package_tag.tag_id AND package.id = package_tag.package_id AND package.state = 'active'
+    GROUP BY tag.name ORDER BY count(tag.name) DESC;
+    """
+
+    cursor.execute(sql)
+    return cursor
+
+
+def create_tag_list(rows):
+    with io.open("invalid_tags.txt", "w+", encoding="utf-8") as f:
+        for row in rows:
+            f.write(row)
+    logger.info('TXT - Written %s lines to invalid_tags.txt', len(rows.split('\n')) - 1)
+
+
+def patch_translator():
+    import ckan.logic.validators
+    def mock_ugettext(value):
+        return unicode(value)
+    ckan.logic.validators._ = mock_ugettext
+
+
+def validate_tag(tag):
+    patch_translator()
+
+    logger.info('VALIDATE - %s', tag)
+
+    tag_json = {
+        'name': tag
+    }
+
+    _, errors = validate(tag_json, default_tags_schema())
+    return errors
+
+
+def main():
+    setup_logging()
+
+    logger.info('Find invalid tags')
+    logger.info('====================================================================')
+
+    txt_rows = ''
+    counter = 0
+    total_count = 0
+
+    for i, tag in enumerate(get_all_tags()):
+        logger.info('%d - %s', i, tag)
+        tag_name, count = tag
+        tag_name_utf8 = tag_name.decode('utf-8')
+        errors = validate_tag(tag_name_utf8)
+        if errors:
+            counter += 1
```

Comment:
If `txt_rows` was stored as a `list` we could use `len(txt_rows)` in place of `counter` in the loop, similarly I think we could get rid of `total_count`.

Reply:
I can get rid of `counter` by using a list but need `total_count` as that accumulates the `count` returned by the validation function